### PR TITLE
feat: new Azure Cosmos DB Module

### LINKED ISSUE #247: [Module Proposal]: Azure Cosmos DB
### Have you checked this module does not already exist in the registry?

Yes

### Why is the module needed?

We require a base module for managing Azure CosmosDB and one is not currently available in the public registry. It will follow the concept of the Azure Container Registry module that was previously created via https://github.com/Azure/bicep-registry-modules/issues/166

Our team is currently scaling up on usage of Bicep and multiple engineers will be contributing many different base modules here, instead of maintaining a private collection of Bicep modules.

### Module path

data/cosmos-db

### Describe the module

This module will allow for provisioning and management of Azure CosmosDB, in line with a similar experience as using Azure Portal/CLI, for example: support for integrating in VNETs via Private Endpoint, databases, etc.

### PR DESCRIPTION
**Pull request by dciborow**
## Description
Closes #247

If you haven't already, read the full [contribution guide](../CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

Enable GitHub Worksflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) as a secret in your forked repository called `PAT`.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [x] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

### CODE REVIEW

**Comment by dciborow** [comment]
@shenglol this is ready for review, but I want to leave the issue open as I'm just trying to provide a foundation for them to extend (thus version = 0.0).

File: modules/data/cosmos-db/README.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
+# Azure Cosmos DB
+
+Bicep module for simplified deployment of Cosmos DB; enables VNet integration and offers flexible configuration options.
+
+## Description
+
+This Bicep Module simplifies the deployment of a Cosmos DB account by providing a reusable set of parameters and resources.
+It allows for the creation of a new Cosmos DB account or use of an existing one, and offers configuration options such as the default consistency level, system-managed failover, and multi-region writes.
+The module also allows for the attachment of the Cosmos DB account to a virtual network, if specified.
```
Comment (shenglol) [inline modules/data/cosmos-db/README.md:9]: Is the statument about vnet attachment true? I didn't find the bicep code regarding that.

Comment (dciborow) [reply modules/data/cosmos-db/README.md:9]: It looks like I lost this in my copying from repo to repo.... let me inspect

Comment (dciborow) [reply modules/data/cosmos-db/README.md:9]:
```suggestion
```

**Comment by vlahane** [comment]
@dciborow Can we use module path as `database/cosmos` or `document-db/cosmos` instead of `data/cosmos-db` ?

**Comment by shenglol** [comment]
Will need to update the MCR manifest to reflect the module group name change.

File: modules/databases/cosmos-db/version.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "$schema": "https://aka.ms/bicep-registry-module-version-file-schema#",
+  "version": "0.0",
```
Comment (dciborow) [inline modules/databases/cosmos-db/version.json:3]:
```suggestion
  "version": "1.0",
```

File: modules/databases/cosmos-db/README.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+```bicep
+param location string = 'eastus'
+param name string = 'mycosmosdb'
+
+param secondaryRegions = [
+  'westus',
+  'centralus'
+]
+
+module cosmosDb 'br/public:data/cosmos-db:0.0.1' = {
```
Comment (dciborow) [inline modules/databases/cosmos-db/README.md:89]:
```suggestion
module cosmosDb 'br/public:storage/cosmos-db:1.0.1' = {
```

**Comment by dciborow** [comment]
> @dciborow Can we use module path as `database/cosmos` or `document-db/cosmos` instead of `data/cosmos-db` ?

After working on some other modules, storage-account, data-explorer, and psql, I'm having a change of heart and want to group all these together under "storage". 

While databases does make sense, I don't think it offers much of a benefit to break these up into its own folder.
And when I am explaining cloud things, I describe services as either "compute" or "storage".

**Comment by shenglol** [comment]
I think it's better to keep database resources and under the group `databases`. The problem with `storage` is that it's a resource provider name, which may lead to confusion.

@alex-frankel what do you think?

**Comment by shenglol** [comment]
> I think I learn `storage` as well actually. We already have this problem with `compute` and `network` as they are both kind of "industry standard" terms.

Got it. In that case, I think we may need to document that the public registry module group names are not RP names, but they are industry standard cloud architecture terms.

File: modules/storage/cosmos-db/version.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "$schema": "https://aka.ms/bicep-registry-module-version-file-schema#",
+  "version": "0.0",
```
Comment (shenglol) [inline modules/storage/cosmos-db/version.json:3]: Should we update the version to `1.0`?

Comment (dciborow) [reply modules/storage/cosmos-db/version.json:3]:
```suggestion
  "version": "1.0",
```

**Review by shenglol — APPROVED** [review]

**Pull request merged**